docs/design/2020-06-24-placement-rules-in-sql.md

```diff
@@ -0,0 +1,391 @@
+
+* ADD: Add new `count` replicas to the table according to `label` and `role` configuration.
+* ALTER: Override the current replica configuration. Old configuration is discarded.
+
+They’re both achieved by executing `ALTER TABLE` statements.
+
+### Adding replicas
+
+Adding new replicas can be done by one or more `ADD PLACEMENT` clauses:
+
+```sql
+ALTER TABLE table_name
+	ADD PLACEMENT LABEL=label ROLE=role COUNT=count,
+	...
+```
+
+This statement indicates TiDB to add replicas for all data, including indexes, of table `table_name`.
+
+`ADD PLACEMENT` is just a part of alter options, just like `ADD COLUMN` or `ADD CONSTRAINT`.
+
+To define multiple roles at once, Multiple `ADD PLACEMENT` clauses can appear in a single `ALTER TABLE` statement, each for one Raft role:
+
+```sql
+ALTER TABLE table_name
+	ADD PLACEMENT LABEL="+zone=sh" ROLE=leader COUNT=1,
+	ADD PLACEMENT LABEL="-zone=sh" ROLE=follower COUNT=2;
+```
+
+This statement indicates PD to schedule the leader to `sh`, and do not schedule followers to `sh`.
+
+`ADD PLACEMENT` can be used to separate data on different data centers, which may help to reduce multi-region latency.
+
+Adding placement can be done through adding a placement rule in PD. The statement must wait until the PD returns a message.
+
+### Altering placement
+
+Altering current placement rules can be done by one or more `ALTER PLACEMENT` clauses:
+
+```sql
+ALTER TABLE table_name
+	ALTER PLACEMENT LABEL=label ROLE=role COUNT=count,
+	...
+```
+
+This statement indicates TiDB to overwrite the current placement rule with the same `role`. It affects all data, including indexes, of table `table_name`.
+
+This kind of operation can be used in adding more replicas for important data, or adding a TiFlash replica.
+
+Altering placement can be done through override the current placement rule by a new one.
+
+Note that when the rules in the statement conflict, an error should occur. For example, an error should be reported when executing this statement:
+
+```sql
+ALTER TABLE test
+	ALTER PLACEMENT LABEL="+zone=sh" ROLE=voter COUNT=3,
+	ALTER PLACEMENT LABEL="+zone=bj" ROLE=voter COUNT=3;
+```
+
+`+zone=sh` conflicts with `+zone=bj`, because they’re defining the same role `voter`. These two constraints can not be satisfied both.
+
+The statement must wait until the PD returns a message.
+
+### Label configuration
+
+`LABEL` in the statement indicates the label constraints. Data is placed on the TiKV nodes whose labels conform to `LABEL` constraint. `LABEL` must be specified in `ADD PLACEMENT` or `ALTER PLACEMENT` clauses.
+
+Parameter `label` should be a string and in such a format: `{+|-}key=value,...`. Prefix `+` indicates that data can only be placed on those TiKV nodes whose labels contain such labels, and `-` indicates that data can’t be placed on those TiKV nodes whose labels contain such labels. `key` here refers to the label name, and `value` is the label value. `key` should be defined in TiKV labels.
```

Comment:
> Prefix `+` indicates that data can only be placed on those TiKV nodes whose labels contain such labels

by `contain`, is it an exact match?

Reply:
No.
If TiKV label is `region=sh, rack=r1`, the rule is `+region=sh`, then it matches. But this is not an exact match.